### PR TITLE
43: adds kafka service discovery

### LINKED ISSUE #43: Implement Cluster Discovery based on Kafka topic
We need to have a mechanic that allows us to discover nodes in the cluster.
For this, we would like to use Kafka (Readpanda)

- [ ] Create a topic `stereodb_cluster_{cluster_id}` with a single partition (for simplicity we can use `auto.create.topics.enable` on kafka settings to create topic automatically)
- [ ] Each node should continue writing into this topic the **Heartbeat** message.
- [ ] The **Heartbeat** message should contain **NodeIP** and should be written with a configurable interval (for now, 5 seconds)
- [ ] Each node should also subscribe on this topic to get information about other nodes in the cluster
- [ ] The topic should be cleaned by Kafka's TTL based cleanup `retention.ms` (for now, 10 minutes)

### PR DESCRIPTION
**Pull request by oleksiikraievyi**
What was done in following PR

+ added creation of heartbeat topic and producing message in configured heartbeat interval
+ added sample consume task which subscribes on all stereodb topics with regex
+ added sample test to check if produce works as expected and marked with Integration category so they will be ignored in CI for now

image
image

### CODE REVIEW

File: src/StereoDB/StereoDb.fs [hunk 1]
```diff
@@ -122,8 +130,44 @@ type internal StereoDb<'TSchema when 'TSchema :> IDbSchema>(logger: ILogger, sch
             with
                 ex -> ()
     }
+
+    let createHeartbeatTopic = task {
+        use adminClient = AdminClientBuilder(AdminClientConfig()).Build()
+        try
+            do! adminClient.CreateTopicsAsync(
```
Comment (oleksiikraievyi) [inline src/StereoDB/StereoDb.fs:145]: TODO @oleksiikraievyi add handling if topic exists + reasonable error messages 
also discuss if we should retry or fail db spinning up

File: src/StereoDB/StereoDb.fs [hunk 2]
```diff
@@ -122,8 +130,44 @@ type internal StereoDb<'TSchema when 'TSchema :> IDbSchema>(logger: ILogger, sch
...
+
+        while _working do
+            try
+                let message = Message<string, string>(
+                    
+                    Value = $"Heartbeat at {DateTime.UtcNow:O}"
+                )
+                producer.ProduceAsync(_heartbeatTopic, message) |> Async.AwaitTask |> ignore
+            with
+            | ex -> logger.Error("Error during sending heartbeat")
```
Comment (oleksiikraievyi) [inline src/StereoDB/StereoDb.fs]: TODO @oleksiikraievyi add reasonable error message

File: src/StereoDB/StereoDb.fs [hunk 3]
```diff
@@ -122,8 +130,44 @@ type internal StereoDb<'TSchema when 'TSchema :> IDbSchema>(logger: ILogger, sch
             with
                 ex -> ()
     }
+
+    let createHeartbeatTopic = task {
+        use adminClient = AdminClientBuilder(AdminClientConfig()).Build()
+        try
+            do! adminClient.CreateTopicsAsync(
+                [TopicSpecification(
+                    Name = _heartbeatTopic
```
Comment (oleksiikraievyi) [inline src/StereoDB/StereoDb.fs]: TODO @oleksiikraievyi add retention.ms to config

File: src/StereoDB/StereoDb.fs [hunk 4]
```diff
@@ -122,8 +130,44 @@ type internal StereoDb<'TSchema when 'TSchema :> IDbSchema>(logger: ILogger, sch
...
+        with 
+        | ex -> logger.Error("Error during creating topic")
+    }
+
+    let startHearbeatTask (config: HeartbeatConfig) = task {
+        do! createHeartbeatTopic
+ 
+        let producerConfig = ProducerConfig(
+            BootstrapServers = config.KafkaBootstrapServers,
+            Acks = Acks.Leader
```
Comment (oleksiikraievyi) [inline src/StereoDB/StereoDb.fs]: TODO @oleksiikraievyi discuss if we should use acks.all or acks.leader

File: src/StereoDB/Infra/Utils.fs [hunk 5]
```diff
@@ -48,4 +48,12 @@ module Seq =
...
\ No newline at end of file
+        MessagePackSerializerOptions.Standard.WithResolver(ContractlessStandardResolverAllowPrivate.Instance)
+
+
+module IPAddress = 
+    open System.Net
+
+    let nodeIp= 
+        let ipHostInfo = Dns.GetHostEntry(Dns.GetHostName())
+        ipHostInfo.AddressList[0].ToString()
```
Comment (oleksiikraievyi) [inline src/StereoDB/Infra/Utils.fs]: TODO @oleksiikraievyi add new lines at the EOF

**Comment by oleksiikraievyi** [comment]
TODO @oleksiikraievyi add integration tests and add category so they will only run if category is set explicitly in `dotnet test`

File: src/StereoDB/StereoDb.fs [hunk 6]
```diff
@@ -122,8 +130,44 @@ type internal StereoDb<'TSchema when 'TSchema :> IDbSchema>(logger: ILogger, sch
...
+        use adminClient = AdminClientBuilder(AdminClientConfig()).Build()
+        try
+            do! adminClient.CreateTopicsAsync(
+                [TopicSpecification(
+                    Name = _heartbeatTopic
+                )],
+                CreateTopicsOptions()
+            )
+        with 
+        | ex -> logger.Error("Error during creating topic")
```
Comment (AntyaDev) [inline src/StereoDB/StereoDb.fs]: you can remove |
```
try
    some code
with
    ex -> ()
```

File: src/StereoDB/StereoDb.fs [hunk 7]
```diff
@@ -122,8 +130,44 @@ type internal StereoDb<'TSchema when 'TSchema :> IDbSchema>(logger: ILogger, sch
...
+            try
+                let message = Message<string, string>(
+                    
+                    Value = $"Heartbeat at {DateTime.UtcNow:O}"
+                )
+                producer.ProduceAsync(_heartbeatTopic, message) |> Async.AwaitTask |> ignore
+            with
+            | ex -> logger.Error("Error during sending heartbeat")
+            
+            do! Async.Sleep(config.HeartbeatInterval)
```
Comment (AntyaDev) [inline src/StereoDB/StereoDb.fs:166]: I would better replace on regular 
`do! Task.Delay`

File: src/StereoDB/StereoDb.fs (hunk 6 above)
Comment (oleksiikraievyi) [reply src/StereoDB/StereoDb.fs]: done

File: tests/StereoDB.Tests/HeartbeatTests.fs [hunk 8]
```diff
@@ -14,25 +14,30 @@ open Tests.TestHelper
 open System
 open System.Threading
 
-[<Fact>]
+[<Fact(Skip = "requires local kafka")>]
```
Comment (kant2002) [inline tests/StereoDB.Tests/HeartbeatTests.fs]: Maybe we can mark with using category, and document command how we can run with test category disabled?

File: src/StereoDB/StereoDb.fs [hunk 9]
```diff
@@ -122,10 +135,53 @@ type internal StereoDb<'TSchema when 'TSchema :> IDbSchema>(logger: ILogger, sch
...
+        use adminClient = AdminClientBuilder(
+            AdminClientConfig(BootstrapServers = config.KafkaBootstrapServers)).Build()
+
+        try
+            // let! existingTopics = adminClient.DescribeTopicsAsync(TopicCollection.OfTopicNames([_heartbeatTopic]))
+            // let heartbeatTopicExists = existingTopics.TopicDescriptions.Exists(
+            //     fun t -> t.Name = _heartbeatTopic && not t.Error.IsError)
+            // if heartbeatTopicExists then
+            //     return! ValueTask.FromResult()
+            let! r = adminClient.CreateTopicsAsync(
```
Comment (AntyaDev) [inline src/StereoDB/StereoDb.fs]: you can use
```let _ = adminClient.CreateTopicsAsync```
I suppose you don't even need 
```r |> ignore```

File: tests/StereoDB.Tests/HeartbeatTests.fs [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+    use cts = new CancellationTokenSource()
+    cts.CancelAfter(TimeSpan.FromSeconds(10)) // wait for heartbeat messages to be sent
+
+    let tcs = TaskCompletionSource()
+    let heartbeatMessages = ConcurrentBag<string>()
+
+    task {
+        kafkaConsumer.Subscribe(settings.HeartbeatConfig.KafkaHeartbeatTopic(settings.ClusterId))
+        while cts.IsCancellationRequested do
```
Comment (AntyaDev) [inline tests/StereoDB.Tests/HeartbeatTests.fs]: maybe 
```while not cts.IsCancellationRequested do```

File: src/StereoDB/StereoDb.fs (hunk 4 above)
Comment (oleksiikraievyi) [reply src/StereoDB/StereoDb.fs]: i guess acks.leader makes more sense in terms of fast insert

File: tests/StereoDB.Tests/HeartbeatTests.fs (hunk 10 above)
Comment (oleksiikraievyi) [reply tests/StereoDB.Tests/HeartbeatTests.fs]: aggree, updated

File: tests/StereoDB.Tests/HeartbeatTests.fs (hunk 8 above)
Comment (oleksiikraievyi) [reply tests/StereoDB.Tests/HeartbeatTests.fs]: yep, agree that were my initial thoughts, updated in recent commit
https://github.com/StereoDB/StereoDB/pull/45#issuecomment-2558068254

File: src/StereoDB/StereoDb.fs (hunk 9 above)
Comment (oleksiikraievyi) [reply src/StereoDB/StereoDb.fs]: rewritten with `do! adminClient.CreateTopicsAsync`

File: src/StereoDB/StereoDb.fs [hunk 11]
```diff
@@ -122,10 +136,70 @@ type internal StereoDb<'TSchema when 'TSchema :> IDbSchema>(logger: ILogger, sch
...
+    }
     
+    let listenHeartbeatTopics (config: HeartbeatConfig) = valueTask {
+        use kafkaConsumer = ConsumerBuilder<string, string>(
+            ConsumerConfig(
+                BootstrapServers = settings.HeartbeatConfig.KafkaBootstrapServers,
+                AutoOffsetReset = AutoOffsetReset.Latest,
+                GroupId = config.KafkaHeartbeatTopicPrefix)).Build()
+
+        kafkaConsumer.Subscribe($"^{config.KafkaHeartbeatTopicPrefix}.+")  // subscribe to all heartbeat topics by regex
```
Comment (AntyaDev) [inline src/StereoDB/StereoDb.fs]: It seems your consumer should listen Heartbeats only from your ClusterId

File: tests/StereoDB.Tests/HeartbeatTests.fs [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+    use! db = StereoDb.init(Schema(), settings)
+    use kafkaConsumer = ConsumerBuilder<string, string>(
+        ConsumerConfig(
+            BootstrapServers = settings.HeartbeatConfig.KafkaBootstrapServers,
+            AutoOffsetReset = AutoOffsetReset.Latest,
+            GroupId = Random.Shared.NextInt64().ToString())).Build()
+
+    use cts = new CancellationTokenSource()
+    cts.CancelAfter(TimeSpan.FromSeconds(10)) // wait for heartbeat messages to be sent
```
Comment (AntyaDev) [inline tests/StereoDB.Tests/HeartbeatTests.fs]: F# is nice, you can write
```cts.CancelAfter(TimeSpan.FromSeconds 10)```

File: tests/StereoDB.Tests/HeartbeatTests.fs [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+    let tcs = TaskCompletionSource()
+    let heartbeatMessages = ConcurrentBag<string>()
+
+    task {
+        kafkaConsumer.Subscribe(settings.HeartbeatConfig.KafkaHeartbeatTopic(settings.ClusterId))
+        try
+            while not cts.IsCancellationRequested do
+                do! Task.Yield()
```
Comment (AntyaDev) [inline tests/StereoDB.Tests/HeartbeatTests.fs]: why do you need Task Yielding for each iteration?
It seems you can use it only once, on top of while

```
do! Task.Yield()
try
    while not cts.IsCancellationRequested do
```

Comment (oleksiikraievyi) [reply tests/StereoDB.Tests/HeartbeatTests.fs:39]: good catch, updated

File: tests/StereoDB.Tests/HeartbeatTests.fs (hunk 12 above)
Comment (oleksiikraievyi) [reply tests/StereoDB.Tests/HeartbeatTests.fs:30]: updated

File: src/StereoDB/StereoDb.fs (hunk 11 above)
Comment (oleksiikraievyi) [reply src/StereoDB/StereoDb.fs]: updated

File: tests/StereoDB.Tests/HeartbeatTests.fs (hunk 12 above)
Comment (oleksiikraievyi) [reply tests/StereoDB.Tests/HeartbeatTests.fs]: updated

File: tests/StereoDB.Tests/HeartbeatTests.fs (hunk 13 above)
Comment (oleksiikraievyi) [reply tests/StereoDB.Tests/HeartbeatTests.fs]: agree, removed, that was initally added to resolve issue with blocking producer and consumer

**Pull request merged**